pkg/client/oci/oci.go

```diff
@@ -0,0 +1,302 @@
+func (r *Repo) List() ([]string, error) {
+	return nil, errors.Errorf("list method is not supported yet")
+}
+
+// getTagManifest returns the manifests of a published tag
+func (r *Repo) getTagManifest(name, version string) (*TagManifest, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), getTimeout)
+	defer cancel()
+
+	u := *r.url
+	// Form API endpoint URL from repo url
+	u.Path = path.Join("v2", u.Path, name, "manifests", version)
+	req, err := http.NewRequestWithContext(ctx, "GET", u.String(), nil)
+	req.Header.Set("Accept", "application/vnd.oci.image.manifest.v1+json")
+
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if r.username != "" && r.password != "" {
+		req.SetBasicAuth(r.username, r.password)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer resp.Body.Close()
+
+	status := resp.StatusCode
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Errorf("unexpected response — %d %q — from %s", status, http.StatusText(status), u.String())
+	}
+	if status != http.StatusOK {
+		return nil, errors.Errorf("unexpected response — %d %q, %s — from %s", status, http.StatusText(status), string(body), u.String())
+	}
+	tm := &TagManifest{}
+	if err := json.Unmarshal(body, tm); err != nil {
+		return nil, err
+	}
+	return tm, nil
+}
+
+// getChartDigest returns the digest of a published chart
+func (r *Repo) getChartDigest(name, version string) (string, error) {
+	tm, err := r.getTagManifest(name, version)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	for _, layer := range tm.Layers {
+		if layer.MediaType == "application/tar+gzip" {
+			return layer.Digest, nil
+		}
+	}
+
+	return "", errors.Errorf("unable to found digest for %s:%s", name, version)
+}
+
+// ListChartVersions lists all versions of a chart
+func (r *Repo) ListChartVersions(name string) ([]string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), getTimeout)
+	defer cancel()
+
+	u := *r.url
+	// Form API endpoint URL from repository URL
+	u.Path = path.Join("v2", u.Path, name, "tags", "list")
+
+	req, err := http.NewRequestWithContext(ctx, "GET", u.String(), nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if r.username != "" && r.password != "" {
+		req.SetBasicAuth(r.username, r.password)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer resp.Body.Close()
+
+	status := resp.StatusCode
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Errorf("unexpected response — %d %q — from %s", status, http.StatusText(status), u.String())
+	}
+	// If status not found 404, it could mean the asset has no release yet
+	if status == http.StatusNotFound {
+		return []string{}, nil
+	}
+	if status != http.StatusOK {
+		return nil, errors.Errorf("unexpected response — %d %q, %s — from %s", status, http.StatusText(status), string(body), u.String())
+	}
```

Comment:
let's use a `switch`. Are these response codes oci-specific or are they harbor-oci-specific?

Reply:
they are valid responses for native OCI:

https://github.com/opencontainers/distribution-spec/blob/master/spec.md#endpoints

Reply:
please add a reference inline so others can check the list of endpoints